grabber/main.go

```diff
@@ -117,44 +119,50 @@ func listener(importer *backend.Backend) {
 	for {
 		select {
 		case <-ticker:
-			latestBlocknumber := getLatestBlockNumber(importer)
-			importer.Lgr.Debug("Getting block in listener", zap.Int64("Block Number", latestBlocknumber))
-			if prevHeader != latestBlocknumber {
-				importer.Lgr.Info("Getting block in listener", zap.Int64("Block Number", latestBlocknumber))
-				block, err := importer.BlockByNumber(latestBlocknumber)
-				if block != nil {
+			latest, err := importer.GetLatestBlockNumber()
+			if err != nil {
+				importer.Lgr.Error("Listener: Failed to get latest block number", zap.Error(err))
+				continue
+			}
+			lgr := importer.Lgr.With(zap.Int64("block", latest))
+			lgr.Debug("Listener: Getting block")
+			if prevHeader != latest {
+				lgr.Info("Listener: Getting block")
+				block, err := importer.BlockByNumber(latest)
+				if err != nil {
+					lgr.Error("Listener: Failed to get block", zap.Error(err))
+				} else if block == nil {
+					lgr.Error("Listener: Block not found")
+				} else {
 					importer.ImportBlock(block)
-					if err != nil {
-						importer.Lgr.Fatal("Listener", zap.Int64("Block Number", latestBlocknumber))
-					}
 					checkAncestors(importer, block.Number().Int64(), 100)
-					prevHeader = latestBlocknumber
+					prevHeader = latest
 				}
 			}
 		}
 	}
 }
 
-func getLatestBlockNumber(importer *backend.Backend) int64 {
-	number, err := importer.GetFirstBlockNumber()
-	if err != nil {
-		importer.Lgr.Fatal("getLatestBlockNumber", zap.Error(err))
-	}
-	return number
-}
-
 // backfill continuously loops over all blocks in reverse order, verifying that all
 // data is loaded and consistent, and reloading as necessary.
-func backfill(importer *backend.Backend, startFrom int64) {
-	blockNumber := startFrom
-	if startFrom <= 0 {
-		blockNumber = getLatestBlockNumber(importer)
-	}
+func backfill(importer *backend.Backend, blockNumber int64) {
 	var hash common.Hash // Expected hash.
 	for {
+		if blockNumber < 0 {
+			// Start over from latest.
+			hash = common.Hash{}
+			var err error
+			blockNumber, err = importer.GetLatestBlockNumber()
+			if err != nil {
+				importer.Lgr.Error("Backfill: Failed to get latest block number", zap.Error(err))
+				time.Sleep(5 * time.Second)
+				continue
```

Comment:
These fatals now continue (w/o incrementing the block number) to retry after waiting for 5 seconds. This means the process will hang on a problem block, instead of shutting everything down and restarting from the latest again. I think that this change is OK, b/c this sort of error is either something temporary that will go through eventually, or something critical that we will need to intervene on.

On the other hand, while the process is 'stuck' on this block, there could be new gaps forming near the 'latest' end of the chain, which it would have been filling in if it had been restarting from latest each time. We could restore that old behavior by jumping to latest on failure, or we could do something more sophisticated like retry a few times, then give up and loop back to latest.